File: src/chain-observers/filters/exchange-transfer-filter.ts
import type { WorkerBee } from "../../bot";
import { isExchange } from "../../utils/known-exchanges";
import { OperationClassifier } from "../classifiers";
import type { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import type { TFilterEvaluationContext } from "../factories/data-evaluation-context";
import { FilterBase } from "./filter-base";

export class ExchangeTransferFilter extends FilterBase {
  public constructor(
    worker: WorkerBee
  ) {
    super(worker);
  }

  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [
      OperationClassifier
    ];
  }

  public async match(data: TFilterEvaluationContext): Promise<boolean> {
    const operations = await data.get(OperationClassifier);

    const transfer = operations.operationsPerType["transfer_operation"];
    const fromSavings = operations.operationsPerType["transfer_from_savings_operation"];
    const escrow = operations.operationsPerType["escrow_transfer_operation"];
    const recurrent = operations.operationsPerType["recurrent_transfer_operation"];

    if(transfer)
      for(const op of transfer)
        if(isExchange(op.operation.from))
          return true;

    if(fromSavings)
      for(const op of fromSavings)
        if(isExchange(op.operation.from))
          return true;

    if(escrow)
      for(const op of escrow)
        if(isExchange(op.operation.from))
          return true;

    if(recurrent)
      for(const op of recurrent)
        if(isExchange(op.operation.from))
          return true;

    return false;
  }
}

File: python/workerbee/chain_observers/filters/exchange_transfer_filter.py
"""ExchangeTransferFilter — matches transfers to/from known exchanges.

Mirrors src/chain-observers/filters/exchange-transfer-filter.ts.

Uses OperationClassifier.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..utils import is_exchange
from .filter_base import FilterBase

if TYPE_CHECKING:
    from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
    from ..factories.data_evaluation_context import DataEvaluationContext


class ExchangeTransferFilter(FilterBase):
    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        from ..classifiers.operation_classifier import OperationClassifier

        return [OperationClassifier]

    async def match(self, data: DataEvaluationContext) -> bool:
        from ..classifiers.operation_classifier import OperationClassifier

        operations = await data.get(OperationClassifier)
        ops = operations["operations_per_type"]

        transfer = ops.get("transfer_operation")
        from_savings = ops.get("transfer_from_savings_operation")
        escrow = ops.get("escrow_transfer_operation")
        recurrent = ops.get("recurrent_transfer_operation")

        if transfer:
            for op in transfer:
                operation = op["operation"]
                if is_exchange(str(operation.get("from", ""))) or is_exchange(str(operation.get("to", ""))):
                    return True

        if from_savings:
            for op in from_savings:
                operation = op["operation"]
                if is_exchange(str(operation.get("from", ""))) or is_exchange(str(operation.get("to", ""))):
                    return True

        if escrow:
            for op in escrow:
                operation = op["operation"]
                if is_exchange(str(operation.get("from", ""))) or is_exchange(str(operation.get("to", ""))):
                    return True

        if recurrent:
            for op in recurrent:
                operation = op["operation"]
                if is_exchange(str(operation.get("from", ""))) or is_exchange(str(operation.get("to", ""))):
                    return True

        return False
